File: backend/src/services/dataStore.ts
/**
 * DataStore - 统一数据访问抽象接口
 *
 * 所有服务通过 DataStore 接口访问数据，不直接依赖具体数据库客户端。
 * 当前实现：PgDataStore（PostgreSQL）
 *
 * Requirements: C1.2, C1.4, I6.17
 */

import type { Pool, PoolClient } from 'pg';

// ─── Types ───────────────────────────────────────────────────────────────────

/**
 * 事务回调接收的客户端接口
 * 在事务内部使用，自动处理 BEGIN/COMMIT/ROLLBACK
 */
export interface DataStoreTransaction {
  query<T>(sql: string, params?: unknown[]): Promise<T[]>;
  queryOne<T>(sql: string, params?: unknown[]): Promise<T | null>;
  execute(sql: string, params?: unknown[]): Promise<{ rowCount: number }>;
}

/**
 * DataStore 抽象接口
 *
 * 提供通用的数据库操作方法，所有服务通过此接口访问数据，
 * 不直接依赖 pg、better-sqlite3 等具体数据库客户端。
 */
export interface DataStore {
  /** 执行查询，返回所有匹配行 */
  query<T>(sql: string, params?: unknown[]): Promise<T[]>;

  /** 执行查询，返回第一行或 null */
  queryOne<T>(sql: string, params?: unknown[]): Promise<T | null>;

  /** 执行写操作（INSERT/UPDATE/DELETE），返回受影响行数 */
  execute(sql: string, params?: unknown[]): Promise<{ rowCount: number }>;

  /** 在事务中执行一组操作，自动处理 BEGIN/COMMIT/ROLLBACK */
  transaction<T>(fn: (tx: DataStoreTransaction) => Promise<T>): Promise<T>;

  /** 获取底层连接池（用于需要直接访问的高级场景） */
  getPool(): Pool;

  /** 健康检查，返回数据库是否可达 */
  healthCheck(): Promise<boolean>;

  /** 关闭连接池，释放所有连接 */
  close(): Promise<void>;
}

File: opsevo-python/src/opsevo/data/datastore.py
"""DataStore abstract base class.

Defines the contract for all data store implementations (PostgreSQL,
mock in-memory, etc.).  Every method is async to support non-blocking I/O.

Requirements: 6.1, 6.2
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Awaitable, Callable, TypeVar

T = TypeVar("T")


class DataStoreTransaction(ABC):
    """Handle exposed inside a transaction callback for multi-statement ops."""

    @abstractmethod
    async def query(self, sql: str, params: tuple | list | None = None) -> list[dict]:
        ...

    @abstractmethod
    async def query_one(self, sql: str, params: tuple | list | None = None) -> dict | None:
        ...

    @abstractmethod
    async def execute(self, sql: str, params: tuple | list | None = None) -> int:
        """Execute a statement and return the number of affected rows."""
        ...


class DataStore(ABC):
    """Abstract data-access interface used throughout the application."""

    @abstractmethod
    async def query(self, sql: str, params: tuple | list | None = None) -> list[dict]:
        """Run a SELECT and return all rows as dicts."""
        ...

    @abstractmethod
    async def query_one(self, sql: str, params: tuple | list | None = None) -> dict | None:
        """Run a SELECT and return the first row, or ``None``."""
        ...

    @abstractmethod
    async def execute(self, sql: str, params: tuple | list | None = None) -> int:
        """Execute a DML statement and return the number of affected rows."""
        ...

    @abstractmethod
    async def transaction(self, fn: Callable[[DataStoreTransaction], Awaitable[T]]) -> T:
        """Run *fn* inside a transaction.

        The callback receives a :class:`DataStoreTransaction` handle.
        If *fn* raises, the transaction is rolled back automatically;
        otherwise it is committed.
        """
        ...

    @abstractmethod
    async def health_check(self) -> bool:
        """Return ``True`` when the underlying store is reachable."""
        ...

    @abstractmethod
    async def close(self) -> None:
        """Release all resources (connection pool, etc.)."""
        ...
